src/main/scala/edu/ie3/simona/model/thermal/ThermalGrid.scala

```diff
@@ -171,122 +184,371 @@ final case class ThermalGrid(
-  ): (ThermalGridState, Option[ThermalThreshold]) =
-    house.zip(state.houseState) match {
-      case Some((thermalHouse, lastHouseState)) =>
-        /* Set thermal power exchange with storage to zero */
-        // TODO: We would need to issue a storage result model here...
-        val updatedStorageState = storage.zip(state.storageState) match {
-          case Some((thermalStorage, storageState)) =>
-            Some(
-              thermalStorage
-                .updateState(
-                  relevantData.currentTick,
-                  zeroKW,
-                  storageState,
-                )
-                ._1
-            )
-          case _ => state.storageState
-        }
+      thermalDemands: ThermalDemandWrapper,
+  ): (ThermalGridState, Option[ThermalThreshold]) = {
+    // TODO: We would need to issue a storage result model here...
+
+    /* Consider the action in the last state */
+    val (qDotHouseLastState, qDotStorageLastState) = thermalGridState match {
+      case ThermalGridState(Some(houseState), Some(storageState)) =>
+        (houseState.qDot, storageState.qDot)
+      case ThermalGridState(Some(houseState), None) => (houseState.qDot, zeroKW)
+      case ThermalGridState(None, Some(storageState)) =>
+        (zeroKW, storageState.qDot)
+      case _ =>
+        throw new InconsistentStateException(
+          "There should be at least a house or a storage state."
+        )
+    }
 
-        val (updatedHouseState, maybeHouseThreshold) =
-          thermalHouse.determineState(
-            relevantData,
-            lastHouseState,
-            lastAmbientTemperature,
-            qDot,
+    if (
+      (qDotHouseLastState > zeroKW && (qDotStorageLastState >= zeroKW)) | (qDotStorageLastState > zeroKW && thermalDemands.heatStorageDemand.hasAdditionalDemand)
+    ) {
+      val (updatedHouseState, thermalHouseThreshold, remainingQDotHouse) =
+        handleInfeedHouse(
+          relevantData,
+          lastAmbientTemperature,
+          thermalGridState,
+          qDotHouseLastState,
+        )
+      val (updatedStorageState, thermalStorageThreshold) =
+        if (
+          qDotStorageLastState >= zeroKW && remainingQDotHouse > qDotStorageLastState
+        ) {
+          handleInfeedStorage(
+            relevantData.currentTick,
+            thermalGridState,
+            remainingQDotHouse,
           )
+        } else {
+          handleInfeedStorage(
+            relevantData.currentTick,
+            thermalGridState,
+            qDotStorageLastState,
+          )
+        }
+
+      val nextThreshold = determineMostRecentThreshold(
+        thermalHouseThreshold,
+        thermalStorageThreshold,
+      )
+      (
+        thermalGridState.copy(
+          houseState = updatedHouseState,
+          storageState = updatedStorageState,
+        ),
+        nextThreshold,
+      )
+    }
+    // Handle edge case where house was heated from storage and HP will be activated in between
```

Comment:
"in between" means from the current tick on?

Reply:
yes, changed it.